Create rudimentary admin interface for reviewing applications

samderanova opened the pull request:
- Allow searching for applicants by name and status
- Clicking a search result displays their responses
- Create accept, reject, and waitlist dummy buttons
    - These buttons will work once the backend is constructed
- Add basic backend route for testing purposes

Commit:
Create draft of admin portal with sample data

- Allow searching for applicants by name and status
- Clicking a search result displays their responses
- Create accept, reject, and waitlist dummy buttons

rileysw reviewed (commented):
Thank you Sam for working on this! For basic functionality (view and change status), this is a great start. There are of course more things we can add and improve later, but applications should be able to work with this. Awesome job!

Notes:
* search by email would also be helpful
* number of applications per status next to the checkboxes would be helpful
* for the next steps in the application process...
     * if we decide to use SendGrid emails instead of manually Gmailing, we may also need a "send accepted/waitlist/rejected email" button but also some sort of fail-safe so admins don't accidentally send anything
     * similarly, an "update application portal" button since the update shouldn't occur on a rolling basis

taesungh commented on src/api/routers/admin.py:
```diff
@@ -0,0 +1,31 @@
+router = APIRouter()
+
+
+async def _get_administrators() -> set[object]:
```
Each user has a role stored in the database, so we can check for a permitted role directly on the user.

Commit:
Optimize checking whether a user is authorized

taesungh commented on src/views/admin/applications/AdminHome.tsx:
```diff
@@ -0,0 +1,161 @@
+	const hasAccess = role === "admin";
+
+	const [userDisplayed, setUserDisplayed] = useState<number>(0);
+	const [sidebarInfo, setSidebarInfo] = useState<SidebarDisplay[]>([
+		{
+			name: "Peter Anteater",
+			status: Status.pending,
+		},
+	]);
```
Suggestion: having a derived state may make the data more difficult to manage. We could compute this data based on a primary reference rather than storing it in state.

taesungh commented on src/views/admin/components/Application/ApplicationSection/ApplicationSection.tsx:
```diff
@@ -0,0 +1,72 @@
+			<h2>{title}</h2>
+			<Row>
+				{propsToShow.map((prop, index) => {
+					const element = generateElement(data, prop);
```
Note: Calling a function component instead of rendering it as an element may lead to unexpected behavior.

taesungh replied:
Resolved in e047f80.

taesungh commented on src/views/admin/components/ApplicationSearch/ApplicationSearch.module.scss:
```diff
@@ -0,0 +1,19 @@
+@import "styles/gallery-theme";
+@import "bootstrap/scss/_functions";
+@import "bootstrap/scss/_variables";
+@import "bootstrap/scss/mixins/_breakpoints";
```
Chore [repeating]: we should use the Bootstrap utilities we have prepared, see comments on #26.

taesungh commented on src/views/admin/applications/AdminHome.tsx:
```diff
@@ -0,0 +1,161 @@
+import { TitleBanner } from "components";
+import Router from "next/router";
+import { useContext, useEffect, useState } from "react";
+import { Col, Row } from "react-bootstrap";
```
Chore [repeating]: When [importing components with React Bootstrap](https://react-bootstrap.github.io/getting-started/introduction#importing-components), we should import individual components.

taesungh commented on src/views/admin/applications/AdminHome.tsx:
```diff
@@ -0,0 +1,161 @@
+						" " +
+						app["application_data"].last_name;
+					let status;
+					switch (app["status"]) {
+						case "ACCEPTED":
+							status = Status.accepted;
+							break;
+						case "REJECTED":
+							status = Status.rejected;
+							break;
+						case "WAITLISTED":
+							status = Status.waitlisted;
+							break;
+						default:
+							status = Status.pending;
+					}
```
Thought: I think we could use a string enum instead of a switch, but I'm not certain of all the implementation details right now.

taesungh commented on src/views/admin/components/ApplicationSearch/ApplicationSearch.tsx:
```diff
@@ -0,0 +1,98 @@
+
+function ApplicationSearch({ data, setUser }: ApplicationSidebarProps) {
+	const [displayedApplicants, setDisplayedApplicants] = useState<number[]>([]);
+	const [showAccepted, setShowAccepted] = useState<boolean>(false);
+	const [showRejected, setShowRejected] = useState<boolean>(false);
+	const [showWaitlisted, setShowWaitlisted] = useState<boolean>(false);
+	const [showPending, setShowPending] = useState<boolean>(true);
```
Suggestion: instead of having several states for the filter, we could store a list of statuses to show.

taesungh commented on src/views/admin/components/ApplicationSearch/SearchResults/SearchResults.tsx:
```diff
@@ -0,0 +1,47 @@
+		<>
+			{results.map((index) => {
+				let statusIcon;
+				switch (data[index].status) {
+					case Status.accepted:
+						statusIcon = checkCircleFill;
+						break;
+					case Status.rejected:
+						statusIcon = xCircleFill;
+						break;
+					default:
+						statusIcon = questionCircleFill;
+				}
```
Suggestion: we could use a regular object as a mapping instead of a switch.

taesungh reviewed (changes requested):
Thanks for starting on this. There are a few things we could clean up for better maintainability. Also, you don't have to address these all yourself, you can ask others for help with some of them.

taesungh committed:
Reorganize admin views and components

- Move admin-related views and components to `admin` directory

taesungh committed:
Enumerate user roles and reorganize user records

taesungh committed:
Create dependency for checking roles on API routes

taesungh committed:
Set up Admin Layout to handle authorization checks

- Handle authorization for all admin routes through the layout
  - Admin pages will have their own sidebar for navigation
- Main Layout copied from WIP on #71

taesungh committed:
Reorganize application data hook with useSWR

- Temporarily hide application search to eventually have list view
- Store uid for current applicant to maintain reference upon mutation
- Pass down full `Applicant` data since review actions will need uid

taesungh committed:
Redesign applicant listing as `ListGroup`

taesungh committed:
Redesign `Application` view as a `Card`

- Enhance type checking with application questions and answers

taesungh committed:
Implement application review endpoint/interface

- Add endpoint for a reviewer to submit a review decision
- Connect review submission with applicant actions

RandyHuynh5815 committed:
Add list of shown state + update component imports

RandyHuynh5815 committed:
Revert isLoggedIn and hasAccess to original state

taesungh committed:
Display submitted reviews on Application

- Parse Applicant records since `reviews` might not exist
- List previously submitted reviews and dates

taesungh committed:
Add unit test for review submission

- Create `UserTestClient` for injecting authentication cookie
- Fix formatting in `ApplicationData.py`

taesungh committed:
Fix self-reviewer check for `ApplicationReviews`

taesungh committed:
Reimplement rudimentary status filters

- Might consider using another library as filters become more complex

taesungh commented on src/views/admin/applications/AdminHome.tsx:
```diff
@@ -0,0 +1,161 @@
+function AdminHome() {
+	const { uid, role } = useContext(UserContext);
+	const isLoggedIn = uid !== null;
+	const hasAccess = role === "admin";
```
`AdminLayout` with role checking added in 653dbef.

taesungh commented on src/views/admin/applications/AdminHome.tsx:
```diff
@@ -0,0 +1,161 @@
+		);
+	} else {
+		Router.push("/");
+	}
```
Access checks elevated to `AdminLayout` as part of 652dbef.

taesungh commented on src/admin/components/Application/Application.tsx:
```diff
@@ -0,0 +1,68 @@
+}
+
+const APPLICATION_SECTIONS: ApplicationSections = {
+	"Personal Information": ["gender", "pronouns", "ethnicity", "is_18_older"],
+	Education: ["university", "education_level", "major"],
+	Experience: ["portfolio_link", "linkedin_link", "resume_url"],
+	"Free Response Questions": [
```
Chore: we forgot to include `is_first_hackathon`

taesungh replied:
Added in c2717dd.

taesungh commented:
Interface will be overhauled by another design framework. For now, reviewers should focus on logic of API endpoints and the integration hook in `useApplicants.tsx`.

rileysw commented on src/admin/components/ApplicantStatus/ApplicantStatus.tsx:
```diff
@@ -0,0 +1,34 @@
+	[Status.rejected]: "rejected",
+	[Status.waitlisted]: "waitlisted",
+	[Status.pending]: "needs review",
+	[Status.reviewed]: "reviewed",
```
Clarification: In what case does status = reviewed but not accepted/rejected/waitlisted?

taesungh replied:
I agree this isn't the cleanest, but I originally had it to allow reviewers to submit their reviews without affecting the actual status of an applicant which would be reflected in the portal (status would be reviewed until released). Maybe we could separate it into applicant status (pending review, decision released) and review status (accepted, rejected, waitlisted), but reviewers originally mentioned that we may need to account for multiple reviews submitted by one person.

rileysw replied:
I see. It looks like applications have what they need so far, and the review process is not exactly flushed out yet, so no additional changes needed.

rileysw commented on src/admin/components/Application/Application.tsx:
```diff
@@ -0,0 +1,68 @@
+	const { _id, application_data, status } = applicant;
+	const { first_name, last_name, submission_time, reviews } = application_data;
+
+	const submittedDate = new Date(submission_time).toDateString();
```
Nitpick: I believe the reference point is still UTC time.

taesungh replied:
Haven't noticed this myself, maybe differences in browser or OS compatibility? Running
```js
new Date("2023-01-12T00:06:00.000000+00:00").toDateString()
```

Should return `'Wed Jan 11 2023'`

rileysw replied:
I think I may have missed that the conversion to local time is handled in the function call.

rileysw commented on src/admin/components/Application/Application.tsx:
```diff
@@ -0,0 +1,68 @@
+}
+
+const APPLICATION_SECTIONS: ApplicationSections = {
+	"Personal Information": ["gender", "pronouns", "ethnicity", "is_18_older"],
```
Clarification: Are we missing email? I saw that first and last name were extracted, but I didn't see email.

taesungh replied:
Initially didn't think it was that important to display email, but do you think we should?

rileysw replied:
In general, I think having quick lookup to contact info would be useful, but since applications is running their own spreadsheets this isn't an issue.

Commit:
Add is_first_hackathon field to admin review portal
